compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -849,7 +850,30 @@ impl<'tcx> TraitRef<'tcx> {
         substs: SubstsRef<'tcx>,
     ) -> ty::TraitRef<'tcx> {
         let defs = tcx.generics_of(trait_id);
-        tcx.mk_trait_ref(trait_id, tcx.mk_substs(&substs[..defs.params.len()]))
+        ty::TraitRef::new(tcx, trait_id, tcx.mk_substs(&substs[..defs.params.len()]))
+    }
+
+    /// Returns a `TraitRef` of the form `P0: Foo<P1..Pn>` where `Pi`
+    /// are the parameters defined on trait.
+    pub fn identity(tcx: TyCtxt<'tcx>, def_id: DefId) -> Binder<'tcx, TraitRef<'tcx>> {
```

Comment:
unhappy with using `ty::Binder` in the return type of this fn but not in the others. How frequently do we rely on the binder and how often can we avoid it? might be nice to remove it if it's not too painful.

Reply:
As it turns out: in 4/8 cases. 39b95c9